modules/library/render/src/main/java/org/geotools/renderer/label/LabelPainter.java

```diff
@@ -605,27 +661,113 @@ public void paintCurvedLabel(LineStringCursor cursor) {
                 }
 
-                // draw halo and label
-                if (labelItem.getTextStyle().getHaloFill() != null) {
-                    configureHalo();
-                    for (int i = 0; i < numGlyphs; i++) {
-                        graphics.setTransform(transforms[i]);
-                        graphics.draw(outlines[i]);
-                    }
-                }
-                configureLabelStyle();
-                for (int i = 0; i < numGlyphs; i++) {
-                    graphics.setTransform(transforms[i]);
-                    graphics.fill(outlines[i]);
-                }
+                allOutlines.add(outlines);
+                allTransforms.add(transforms);
 
                 // take into account eventual spaces at the end of the glyph
                 cursor.moveTo(start + glyphVector.getGlyphPosition(numGlyphs).getX());
+            }
+
+            // draw halo and label
+            if (labelItem.getTextStyle().getHaloFill() != null) {
+                configureHalo();
+                if (labelItem.isTextUnderlineEnabled()) {
+                    // we need to draw the underline halo
+                    drawCurvedUnderline(line, cursor, startOrdinate, true);
+                }
+                drawOrFillOutlines(allOutlines, allTransforms, false);
+            }
+            graphics.setTransform(oldTransform);
+            configureLabelStyle();
+            if (labelItem.isTextUnderlineEnabled()) {
+                // we need to draw the underline
+                drawCurvedUnderline(line, cursor, startOrdinate, false);
+            }
+            drawOrFillOutlines(allOutlines, allTransforms, true);
+        } finally {
+            graphics.setTransform(oldTransform);
+        }
+    }
+
+    /**
+     * Helper method that will draw the underline of a curved label using the context of the cursor.
+     */
+    private void drawCurvedUnderline(LineInfo line, LineStringCursor cursor, double startOrdinate, boolean drawingHalo) {
+        // extracting label first line component and compute is metrics
+        LineComponent component = line.getComponents().get(0);
+        LineMetrics metrics = computeLineMetricsIfNeeded(component);
+        // the cursor is in the last char of the label
+        double endOrdinate = cursor.getCurrentOrdinate();
+        // compute the advance based on the first char of the label
+        GlyphVector glyphVector = line.getComponents().get(0).getGlyphVector();
+        double advance = glyphVector.getGlyphMetrics(0).getAdvance() * 0.5f;
+        // extract from the linestring the portion associated with the layer
+        LineString labelLineString = cursor.getSubLineString(startOrdinate - advance, endOrdinate - advance);
+        // compute the underline linestring
+        LiteShape underlineLineString = computeCurvedUnderline(labelLineString, metrics);
+        if (drawingHalo) {
+            // draw the underline halo
+            graphics.draw(underlineLineString);
```

Comment:
Regarding the first question:
Both the methods when drawing the halo are invoked in a situation where configureHalo() has already be invoked, this was on propose to avoid resetting all the options again, but this is not very clear adding a comment about this.